proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,710 @@
+```wasm
+(module
+  (import "Win32" "ReadFile" (func (param i32 i32 i32 i32) (result i32)))
+  (import "Win32" "WriteFile" (func (param i32 i32 i32 i32) (result i32)))
+  (exports $WasiFile)
+)
+```
+
+Lastly, just as the current [text format conventions recommend `.wat`][WAT] as
+the extension of a file that contains a module definition, this proposal
+includes a new recommendation for text files containing a bare module or
+instance type be suffixed `.wit`. `.wit` files can be used as part of the
+toolchain ecosystem for describing a module's interface without including its
+definition. For example, this can be used to generate compatible source-language
+declarations.
+
+Module and Instance types can also be used to describe the types of imports:
+
+
+### Instance Imports
+
+Just as a function, memory, table or global can be imported by specifying a
+function, memory, table or global type, modules or instances can be imported
+by specifying a module or instance type.
+
+A single instance import is practically equivalent to individually importing
+all of its exports. Indeed, instances have no distinct identity: instances are
+just [immutable tuples][`moduleinst`] containing the *addresses* of mutable
+things with identity (that could be independently imported) like memories and
+tables. For example, the instance import in this module:
+```wasm
+(module
+  (import "i" (instance $i
+    (export "f1" (func $f1))
+    (export "f2" (func $f2 (param i32)))
+  ))
+  (func (export "run")
+    call $i.$f1
+  )
+)
+```
+is practically equivalent to individually importing the instance's exports:
+```wasm
+(module
+  (import "i" "f1" (func $f1))
+  (import "i" "f2" (func $f2 (param i32)))
+  (func (export "run")
+    call $f1
+  )
+)
+```
+In particular, the exports of an instance import are injected into the
+appropriate [index spaces] of the containing instance, allowing them to be
+directly accessed by `call`, `i32.load`, `table.get`, etc. The `$i.$f` textual
+notation is new, but still encodes in the binary format as a single integer
+immediate. The reason for the preceding `$i.` is that a single instance type can
+be factored out into a type definition and reused multiple times:
+```wasm
+(module
+  (type $I (instance
+    (export "f" (func $f))
+    (export "g" (func $g))
+  ))
+  (import "i1" (instance $i1 (type $I)))
+  (import "i2" (instance $i2 (type $I)))
+  (func (export "run")
+    call $i1.$f
+    call $i1.$g
+    call $i2.$f
+    call $i2.$g
+  )
+)
+```
+In the following sections, the utility of instance imports will be more clear as
+a concise way to pass around whole instances instead of all their exports
+individually. But for now, one immediate benefit of instance imports is that
+they allow for module and export names to be maximally factored out for both
+text- and binary-format reuse.
+
+
+### Module Imports and Nested Instances
+
+Symmetric to instances, modules can be imported via module type. Once a module
```

Comment:
IIUC, instance imports can be losslessly desugared to existing syntax, while module imports are really a new feature. Maybe it would be clearer not to start out by mentioning symmetry to instance imports.

Reply:
Well, as discussed in the preceding comment, instance imports would be fundamentally new too.

Reply:
Actually, the symmetry I'm referring to here is really with *all* imports (the point being: when you import a thing, you give the type of that thing), so I'll reword to that.